# 01. Importing Libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Set Path

In [5]:
# path
path = r'/Users/renatabatista/Other Docs/Germany/CareerFoundry/Data Imersion/4. Python/14-04-2025 Instacart Basket Analysis'

# 03. Importing datasets

In [7]:
# Products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)

In [8]:
# Orders data - orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_wrangled.csv'), index_col = False)

# 04. Data Consistency Check

In [10]:
# Reviewing Descriptive Statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [11]:
df_ords.drop(columns=['Unnamed: 0'], inplace =True)

In [12]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### A. Mixed-Type Data

###### Instacart data has already undergone all these data-prep checks, and you know there aren’t any mixed-type columns.Let’s create a small test dataframe for you to work with, so you can practice. To make this new test dataframe, execute the following code:

In [15]:
# Create a dataframe
df_test = pd.DataFrame()

In [16]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [17]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [18]:
# Function to check whether a dataframe contains any mixed-type columns:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [19]:
# Converting the mixed column data type to String as most of the data in the mixed column is of a string.
df_test['mix'] = df_test['mix'].astype('str')

In [20]:
df_test['mix'].dtype

dtype('O')

#### B. Missing Values

In [22]:
# Locate missing values and sum the total missing
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
# Create a subset dataframe containing the 16 missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [25]:
# Explanatory code you can have the MEAN. Mean as a statistical measure can also be higly influenced by extreme values, which could lead to innacurate imputions.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [26]:
# To create an impute for missing values you can also use MEDIAN calculation

#### * Impute Missing Values

###### 1.Create a new variable that acts like a flag based on the missing value
###### 2.Impute the value with the MEAN or MEDIAN of the column (if the variable is numeric)
######   - MEAN
######         df['column with missings'].fillna(mean value, inplace=True)
######   -MEDIAN
######         df['column with missings'].fillna(median value, inplace=True)
###### 3.Remove or filter out the missing data

In [29]:
# At this case the data-type missing is a STRING

###### 1. Remove the missing values entirely 
###### or
###### 2. Filter the non-missing values into a subset dataframe

In [31]:
# Compare the number of rows in the current dataframe with the number in your subset once the missing rows are removed
df_prods.shape

(49693, 5)

In [32]:
# Create a new dataframe, setting the 'isnull()' condition to False so that only non-missing values are present
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [33]:
# Check the number of rows again to make sure it descreased
df_prods_clean.shape

(49677, 5)

###### An alternative way to drop all missing values:
#### df_prods.dropna(inplace = True)
###### An alternative way to drop only NaNs missing values from a particular column:
#### df_prods.dropna(subset = ['product_name'], inplace = True)
######
###### - In these cases, rather than creating a new dataframe, you're overwriting df_prods with a new version of df_prods that doesn't contain missing values
###### - This is done using the 'inplace = True' function, which overwrites the dataframe
###### - If you don't specift an 'inplace' argument, the default is False which only returns a view of the changed dataframe

#### C. Duplicates

In [36]:
# Finding Full duplicates within your dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [37]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [38]:
# Lets check the current number of rows 
df_prods_clean.shape

(49677, 5)

###### Addressing Duplicates
###### Once the duplicates are identified, you should remove them
###### Now create a new dataframe without the duplicates 

In [40]:
# Deleting duplicates and creating a CLEAN dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [41]:
df_prods_clean_no_dups.shape

(49672, 5)

## 05. Tidying Up and Exporting Changes

In [43]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

##  06. Data Consistency Check - df_ords

In [45]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### order_id / user_id / order_number
###### should be considered as STRING values instead of numerical values
#### order_number
###### 100 shows the max items ordered number
#### orders_day_of_week
###### Has min = 0 and max = 6. Meaning the seven numbers (0 indexed for Sunday to 6 Saturday) representing the 7days of the week.
#### order_hour_of_day
###### Has a min = 0 and max = 23. Meaning the twenty-four numbers (0 to 23) representing 24hrs of the day.
#### days_since_prior_order
###### Count is less than all other rows inndicating missing values.
###### It also has a measure of zero indicating that some orders could be places within the same day.
###### The max value is 30, which may indicate that orders are capped at 30 days or 1 month.

#### A. Mixed-Type Data

In [48]:
# Function to check whether a dataframe contains any mixed-type columns:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [49]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


#### B. Missing values

In [51]:
# Locate missing values and sum the total missing
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

###### At the column 'days_since_prior_order' is missing 206209 values
###### Maybe the 1st order from the new customers don't have a value
#### However, this is relevant information

#### C. Addressing missing values

In [54]:
# Create a subset dataframe containing the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [55]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

##### Now we need to check if the 'order_number' values = 1 are missing values in the 'days_since_prior_order' column

In [57]:
# Filter the df if column 'order_number' = 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [58]:
# Check if all of these have missing values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [59]:
# Print the result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing) 

All first orders have missing 'days_since_prior_order': True


##### Every row with 'order_number' = 1 has a missing value in the'days_since_prior_order' column

In [61]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [62]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 7 columns]

In [63]:
# re-check the df for accuracy
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

#### D. Duplicate Values

In [124]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [126]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

##### The dataframe created to check duplicates has returned empty, meaning there are no duplicates in the data set

#### E. Export dataframe

In [130]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [132]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))